examples/computer_vision/vision_transformer/det_callback.py

```diff
@@ -0,0 +1,150 @@
+        import determined
+        self._det = determined
+        distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
+
+        self.filter_metrics = filter_metrics
+        self.user_checkpoint_metadata = checkpoint_metadata
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_eval_metrics = None
+        self.searcher_metric = self._det.get_cluster_info().trial._config['searcher']['metric']
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+    def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, model=None, logs=None,
+               **kwargs):
+        if state.is_world_process_zero:
+            metrics = self.process_log(logs)
+            if 'loss' in metrics:
+                self.core_context.train.report_training_metrics(steps_completed=state.global_step, metrics=metrics)
+            elif 'eval_loss' in metrics:
+                self.core_context.train.report_validation_metrics(steps_completed=state.global_step, metrics=metrics)
+                self.last_eval_metrics = metrics
+            else:
+                logging.warning(
+                    f"Metrics not reported: unknown metric type in {logs}.")
+
+    def process_log(self, logs: typing.Dict) -> typing.Dict:
+        metrics = logs
+
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics
+
+    def on_save(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        info = self._det.get_cluster_info()
+        assert info is not None
+        if state.is_world_process_zero:
+            save_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+            det_checkpoint_metadata = {"steps_completed": state.global_step, "trial_id": info.trial.trial_id}
+
+            if self.tokenizer_options is not None:
+                det_checkpoint_metadata['tokenizer_options'] = self.tokenizer_options
+
+            if self.tokenizer is not None:
+                self.tokenizer.save_pretrained(os.path.join(save_path, "tokenizer"))
+
+            if self.user_checkpoint_metadata is not None:
+                self._on_save_user_data(save_path)
+
+            self.core_context.checkpoint.upload(save_path, det_checkpoint_metadata)
+
+        if self.core_context.preempt.should_preempt():
+            raise Exception("Process preempted / killed")
```

Comment:
AFAIK, that's the safest place to do preemption, since Trainer completed saving its data, and determined uploaded directory with Trainer data, tokenizer and other user-specific objects to persistent storage.

In Trainer API loop on_epoch_end/on_step end happens before on_save, so we don't want to preempt there.